### PR TITLE
Add video block size validation in the Media Library

### LINKED ISSUE #2560: Display a notice if a video in the video block is over 1 MB per second
* Splitting off a requirement from #2509 into this issue
* There should be a warning when adding a video to the Video block that is over 1 MB per second:
video-block

As @swissspidy mentioned:
>A suggestion of mine was to also show the warnings when selecting the video in the media modal, just like we do for the featured image.

featured-im

### PR DESCRIPTION
**Pull request by kienstra**
* In the Media Library, on selecting a video with the wrong size, this displays a notice
* Applies to the Media Library for the Video block and 'Background Media' 
* Doesn't disable the 'Select' button
* Branched off #2737, as that refactored the Media Library logic
* If/when that PR is merged, I'll probably rebase this PR to `develop` and target this PR to `develop`

in-video

Closes #2560

### CODE REVIEW

File: assets/src/common/helpers/test/getVideoBytesPerSecond.js [hunk 1]
```diff
@@ -0,0 +1,37 @@
+/**
```
Comment (kienstra) [inline assets/src/common/helpers/test/getVideoBytesPerSecond.js]: This is mainly moved from `stories-editor/helpers/test/getVideoBytesPerSecond.js`, without many changes. Likewise for `isVideoSizeExcessive.js`.

**Comment by kienstra** [comment]
**Steps To Test**

1. Create a new story
2. For the 'Background Media' of a page, upload or select from the 'Media Library' tab a video with more than 1 MB per second, [like this](https://drive.google.com/file/d/1gM3_6iarZcrjeXo2lOePVSPxlUEVKiiU/view?usp=sharing)
3. Excepted: there is a notice like:

>A video size of less than 1 MB per second is recommended. The selected video is 10 MB per second.

notice-appears

**Comment by kienstra** [comment]
This isn't quite ready for review, as it's still targeting the `fix/upload-file-type` branch. If and when that branch is merged into `develop`, I'll probably rebase this to `develop` and target that branch.

**Comment by westonruter** [comment]
@kienstra I merged `develop` and changes the base branch, but then I got confused, so I reset it back!

**Comment by kienstra** [comment]
OK, all good!

**Comment by swissspidy** [comment]
Just rebased the branch and fixed merge conflicts

**Comment by kienstra** [comment]
>Just rebased the branch and fixed merge conflicts

Nice, thanks!

**Comment by kienstra** [comment]
**Thanks for unblocking this**

Hi @swissspidy,
Thanks a lot for rebasing this branch and handling the merge conflicts. 

It's Saturday, so no hurry 😄 

But I just tested this, and it should be ready for review.

**Comment by miina** [comment]
That's not related to the Media Library message directly, however, should there be a warning message outside of Media Library as well, as it is for the Background Media?
1. When inserting the video from URL then there is no warning displayed.
2. When first selecting a video -> closing the Media Library -> clicking the Edit Video icon -> Choosing "Upload" the Media Library won't open and thus the warning isn't displayed.

Thoughts?

**Comment by swissspidy** [comment]
> When inserting the video from URL then there is no warning displayed.

I don't think we can check the video length _and_ file size without downloading the video.

**Comment by miina** [comment]
> I don't think we can check the video length and file size without downloading the video.

I was actually testing with local URL-s where we'd probably have the image data but it's true that it's much more likely that the URL would be used for external images where getting the size and length needs downloading.

I guess the only place where the warning is missing then is when a previous video is changed via  "Edit Video" and "Upload".

**Comment by westonruter** [comment]
> I don't think we can check the video length _and_ file size without downloading the video.

One option would be to do a `HEAD` request for the video URL. That would at least get the file size in `Content-Length`.

**Comment by swissspidy** [comment]
A `HEAD` request is also what came to find. For the video length it might be actually quite easy:

When inserting a video using a URL, we can just access the `<video>` element's duration property, which gives the duration in seconds.

**Comment by westonruter** [comment]
> When inserting a video using a URL, we can just access the `<video>` element's duration property, which gives the duration in seconds.

Great idea. I believe this just means waiting for the [`loadedmetadata`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLMediaElement/loadedmetadata_event) event.

**Comment by kienstra** [comment]
I merged in `develop`, as there is a new custom Video block edit component from #2799 that was just merged.

**Comment by swissspidy** [comment]
Added a warning now when pasting a video URL:

Screenshot 2019-07-16 at 11 26 18

**Comment by miina** [comment]
I'm seeing `0` at the place of the notice if there is no error (see below "Replace Image")

Screenshot 2019-07-16 at 13 05 35

**Comment by miina** [comment]
Likely not the most common use case but looks like displaying the warning based on URL stops showing when I do the following:
- Add a video from URL that exceeds the limits.
- Click "Edit Video" icon.
- Choose "Insert from URL".
- Don't change anything, leave the same URL, submit.

**Comment by miina** [comment]
> I guess the only place where the warning is missing then is when a previous video is changed via "Edit Video" and "Upload".

Actually, there is no need to change a previous video, just using "Upload" directly and choosing a video exceeding the limits won't show the warning since the Media Library doesn't open then.

**Comment by swissspidy** [comment]
> Likely not the most common use case but looks like displaying the warning based on URL stops showing when I do the following

Cannot reproduce this.

> I'm seeing 0 at the place of the notice if there is no error (see below "Replace Image")

Cannot reproduce either.

> Since 4 is used for every test, maybe for readability we could use a general descriptive variable for the seconds

Made the tests much more readable now in 3ce5e76b6ee8868c814a0a248e3ac0b7e679f8df.

**Comment by swissspidy** [comment]
> Actually, there is no need to change a previous video, just using "Upload" directly and choosing a video exceeding the limits won't show the warning since the Media Library doesn't open then.

I guess we can then just show the warning in the sidebar.

**Comment by swissspidy** [comment]
@miina Mind checking again?

**Comment by miina** [comment]
Awesome, the issues mentioned before are gone now!

One more oddity that I just saw, the same video shows different values in the Media Library vs in the sidebar.
Sidebar:
Screenshot 2019-07-16 at 15 10 24

Media Library:
Screenshot 2019-07-16 at 15 10 13

Thoughts?

**Comment by swissspidy** [comment]
> One more oddity that I just saw, the same video shows different values in the Media Library vs in the sidebar.

Interesting. Will try to reproduce.

**Comment by miina** [comment]
> Interesting. Will try to reproduce.

Note that I was using [this video](https://drive.google.com/file/d/1gM3_6iarZcrjeXo2lOePVSPxlUEVKiiU/view?usp=sharing).

**Comment by swissspidy** [comment]
Looks like it was a rounding issue.

**Pull request merged**